java/ql/lib/semmle/code/java/os/OSCheck.qll

```diff
@@ -0,0 +1,93 @@
+/**
+ * Provides classes and predicates for guards that check for the current OS.
+ */
+
+import java
+import semmle.code.java.controlflow.Guards
+private import semmle.code.java.frameworks.apache.Lang
+private import semmle.code.java.dataflow.DataFlow
+private import semmle.code.java.StringCheck
+
+/**
+ * A complimentatry guard that checks if the current platform is Windows.
+ */
+abstract class IsWindowsGuard extends Guard { }
+
+/**
+ * A complimentatry guard that checks if the current platform is unix or unix-like.
+ */
+abstract class IsUnixGuard extends Guard { }
+
+/**
+ * Holds when the MethodAccess is a call to check the current OS using either the upper case `osUpperCase` or lower case `osLowerCase` string constants.
+ */
+bindingset[osUpperCase, osLowerCase]
+private predicate isOsFromSystemProp(MethodAccess ma, string osUpperCase, string osLowerCase) {
```

Comment:
Only one of `osUpperCase` or `osLowerCase` is bound in disjuncts 1 and 2 respectively. Thanks to the bindingset annotation this doesn't try to compute the product of the bound parameter and all possible strings, but it does compute the product with whatever strings are specified at the various invocation sites.

In general you should make sure a parameter is bound in all disjuncts.

Reply:
Sorry, the above is going a bit over my head, can you try the ELI5 version? Or something maybe a bit higher than that, but still simpler 😆

Reply:
Take for example this predicate:

```
bindingset[x,y]
predicate xOrYIsOne(int x, int y) {
  x = 1 or y = 1
}
```

This is made of two disjuncts, each of which only binds one of x or y.

Since these are `int`s, the only way this will compile at all is with a bindingset annotation. Let's say the use site constrains them to be either 1 or 2.

Then this will be computed by evaluating `x = 1`, then `y = 1`, then adding the results together.

That means we'll get `(1, 1), (1, 2)` + `(1, 1), (2, 1)`

If the surrounding environment constrained them to be between 1 and 1000 then we'd get (1, 1), (1, 2) ... (1, 1000) and similarly the other way around.

If they were naturally finite types, like database types, then we could remove the bindingset annotation and we'd get  a hit for every possible value of the unbound parameter.

In the example given here, the use sites limit us to a small range of possible parameter values so the result duplication isn't too bad, but it's bad practice in general to have a disjunct that leaves a parameter unspecified like this.

Instead you could do:

```
bindingset[osUpperCase, osLowerCase]
private predicate isOsFromSystemProp(MethodAccess ma, string osString) {
  exists(MethodAccessSystemGetProperty sgpMa, Expr sgpFlowsToExpr |
    sgpMa.hasCompileTimeConstantGetPropertyName("os.name")
  |
    DataFlow::localExprFlow(sgpMa, sgpFlowsToExpr) and // Call from System.getProperty to some partial match method
    ma.getAnArgument().(CompileTimeConstantExpr).getStringValue().toLowerCase().matches(osString) and
    (
      sgpFlowsToExpr = ma.getQualifier() 
      or
      exists(MethodAccess caseChangeMa |
        caseChangeMa.getMethod() =
          any(Method m | m.getDeclaringType() instanceof TypeString and m.hasName(["toLowerCase", "toUpperCase"]))
        |
        sgpFlowsToExpr = caseChangeMa.getQualifier() and // Call from System.getProperty to case-switching method
        DataFlow::localExprFlow(caseChangeMa, ma.getQualifier()) and // Call from case-switching method to some partial match method
    )
  ) and
  isStringPartialMatch(ma)
}
```

Reply:
That also regularises the use of the case-switching methods. `osString` should be lower-case in this implementation.